File: lib/day06.ex
defmodule Day6 do
  use Day, day: 6

  @test_input """
  ....#.....
  .........#
  ..........
  ..#.......
  .......#..
  ..........
  .#..^.....
  ........#.
  #.........
  ......#...
  """

  @doc """
      iex> Day6.part1(Day6.test_input())
      41

      iex> Day6.part1(Day6.input())
      5199
  """
  def part1(input) do
    matrix =
      input
      |> String.split("\n", trim: true)
      |> Enum.map(&String.split(&1, "", trim: true))

    num_rows = length(matrix)
    num_cols = length(hd(matrix))

    obstacles =
      for {row, i} <- Enum.with_index(matrix),
          {entry, j} <- Enum.with_index(row),
          entry == "#" or entry == "^",
          reduce: %{} do
        acc ->
          if entry == "#" do
            Map.put(acc, {i, j}, entry)
          else
            Map.put(acc, "start", {i, j})
          end
      end

    start = obstacles["start"]

    visited = loop1(start, :up, obstacles, MapSet.new([start]), num_rows, num_cols)

    MapSet.size(visited)
  end

  defp loop1(position, direction, obstacles, visited, num_rows, num_cols) do
    visited = MapSet.put(visited, position)
    {i, j} = next_position = move(position, direction)

    cond do
      Map.has_key?(obstacles, next_position) ->
        new_direction = turn_right(direction)
        loop1(position, new_direction, obstacles, visited, num_rows, num_cols)

      i < 0 or i >= num_rows or j < 0 or j >= num_cols ->
        visited

      true ->
        loop1(next_position, direction, obstacles, visited, num_rows, num_cols)
    end
  end

  defp move({i, j}, :up), do: {i - 1, j}
  defp move({i, j}, :down), do: {i + 1, j}
  defp move({i, j}, :left), do: {i, j - 1}
  defp move({i, j}, :right), do: {i, j + 1}

  defp turn_right(:up), do: :right
  defp turn_right(:right), do: :down
  defp turn_right(:down), do: :left
  defp turn_right(:left), do: :up

  @doc """
      iex> Day6.part2(Day6.test_input())
      6

      # iex> Day6.part2(Day6.input())
      # 1915
  """
  def part2(input) do
    matrix =
      input
      |> String.split("\n", trim: true)
      |> Enum.map(&String.split(&1, "", trim: true))

    num_rows = length(matrix)
    num_cols = length(hd(matrix))

    obstacles =
      for {row, i} <- Enum.with_index(matrix),
          {entry, j} <- Enum.with_index(row),
          entry == "#" or entry == "^",
          reduce: %{} do
        acc ->
          if entry == "#" do
            Map.put(acc, {i, j}, entry)
          else
            Map.put(acc, "start", {i, j})
          end
      end

    start = obstacles["start"]

    for i <- 0..(num_rows - 1),
        j <- 0..(num_cols - 1),
        results_in_cycle?(i, j, start, obstacles, num_rows, num_cols),
        reduce: 0 do
      acc -> acc + 1
    end
  end

  defp results_in_cycle?(i, j, start, obstacles, num_rows, num_cols) do
    obstacles = Map.put(obstacles, {i, j}, "#")

    loop2(start, :up, obstacles, MapSet.new(), num_rows, num_cols)
  end

  defp loop2(position, direction, obstacles, prev_visited, num_rows, num_cols) do
    visited = MapSet.put(prev_visited, {position, direction})
    {i, j} = next_position = move(position, direction)

    cond do
      prev_visited == visited ->
        true

      Map.has_key?(obstacles, next_position) ->
        new_direction = turn_right(direction)
        loop2(position, new_direction, obstacles, visited, num_rows, num_cols)

      i < 0 or i >= num_rows or j < 0 or j >= num_cols ->
        false

      true ->
        loop2(next_position, direction, obstacles, visited, num_rows, num_cols)
    end
  end
end

File: lib/2024/zig_solutions/day06.zig
const beam = @import("beam");
const std = @import("std");

const Direction = enum {
    up,
    down,
    left,
    right,
};

pub fn solve_part1(input: []const u8) i64 {
    var gpa = beam.make_general_purpose_allocator_instance();
    const allocator = gpa.allocator();

    var matrix = std.ArrayList(std.ArrayList(u8)).init(allocator);

    var iter = std.mem.split(u8, input, "\n");
    var pos: [2]usize = undefined;

    var i: usize = 0;
    while (iter.next()) |line| {
        var row = std.ArrayList(u8).init(allocator);
        var j: usize = 0;
        for (line) |c| {
            if (c == '^') {
                pos = .{ i, j };
            }
            row.append(c) catch unreachable;
            j += 1;
        }
        if (row.items.len > 0) {
            matrix.append(row) catch unreachable;
        }
        i += 1;
    }

    const num_rows = matrix.items.len;
    const num_cols: usize = matrix.items[0].items.len;

    var direction = Direction.up;

    var visited = std.AutoHashMap([2]usize, void).init(allocator);

    while (true) {
        const next_pos = move(pos, direction) catch break;

        if (next_pos[0] < 0 or next_pos[0] >= num_rows or next_pos[1] < 0 or next_pos[1] >= num_cols) {
            break;
        }

        if (matrix.items[next_pos[0]].items[next_pos[1]] == '#') {
            direction = turn_right(direction);
        } else {
            matrix.items[pos[0]].items[pos[1]] = '.';
            pos = next_pos;
            var c: u8 = '^';
            if (direction == .down) {
                c = 'v';
            } else if (direction == .left) {
                c = '<';
            } else if (direction == .right) {
                c = '>';
            }
            matrix.items[pos[0]].items[pos[1]] = c;
        }

        visited.put(pos, {}) catch unreachable;
    }

    return visited.count();
}

fn print_matrix(matrix: std.ArrayList(std.ArrayList(u8))) void {
    for (matrix.items) |row| {
        for (row.items) |c| {
            std.debug.print("{c}", .{c});
        }
        std.debug.print("\r\n", .{});
    }
}

fn move(pos: [2]usize, direction: Direction) ![2]usize {
    switch (direction) {
        .up => return if (pos[0] == 0) error.OutOfBounds else .{ pos[0] - 1, pos[1] },
        .down => return .{ pos[0] + 1, pos[1] },
        .left => return if (pos[1] == 0) error.OutOfBounds else .{ pos[0], pos[1] - 1 },
        .right => return .{ pos[0], pos[1] + 1 },
    }
}

fn turn_right(direction: Direction) Direction {
    return switch (direction) {
        .up => .right,
        .right => .down,
        .down => .left,
        .left => .up,
    };
}

pub fn solve_part2(input: []const u8) i64 {
    var gpa = beam.make_general_purpose_allocator_instance();
    const allocator = gpa.allocator();

    var matrix = std.ArrayList(std.ArrayList(u8)).init(allocator);

    var iter = std.mem.split(u8, input, "\n");
    var start: [2]usize = undefined;

    var r: usize = 0;
    while (iter.next()) |line| {
        var row = std.ArrayList(u8).init(allocator);
        var j: usize = 0;
        for (line) |c| {
            if (c == '^') {
                start = .{ r, j };
            }
            row.append(c) catch unreachable;
            j += 1;
        }
        if (row.items.len > 0) {
            matrix.append(row) catch unreachable;
        }
        r += 1;
    }

    const num_rows = matrix.items.len;
    const num_cols: usize = matrix.items[0].items.len;
    var count: i64 = 0;

    for (0..num_rows) |i| {
        for (0..num_cols) |j| {
            var pos = start;
            var direction = Direction.up;
            var visited = std.AutoHashMap(struct { pos: [2]usize, direction: Direction }, bool).init(allocator);
            var results_in_cycle = false;
            while (true) {
                const next_pos = move(pos, direction) catch break;
                if (next_pos[0] < 0 or next_pos[0] >= num_rows or next_pos[1] < 0 or next_pos[1] >= num_cols) {
                    results_in_cycle = false;
                    break;
                }

                if ((next_pos[0] == i and next_pos[1] == j) or matrix.items[next_pos[0]].items[next_pos[1]] == '#') {
                    direction = turn_right(direction);
                } else {
                    matrix.items[pos[0]].items[pos[1]] = '.';
                    pos = next_pos;
                    var c: u8 = '^';
                    if (direction == .down) {
                        c = 'v';
                    } else if (direction == .left) {
                        c = '<';
                    } else if (direction == .right) {
                        c = '>';
                    }
                    matrix.items[pos[0]].items[pos[1]] = c;
                }

                const key = .{ .pos = pos, .direction = direction };

                if (visited.get(key)) |_| {
                    results_in_cycle = true;
                    break;
                }

                visited.put(key, true) catch unreachable;
            }

            if (results_in_cycle) {
                count += 1;
            }
        }
    }

    return count;
}
